capture.go

```diff
@@ -68,63 +86,191 @@ func WithDir(dir string) Option {
+// in the case of failure to run.
+func (c *Capture) Start(ctx context.Context, name string, args ...string) (*Pipes, error) {
 	c.Name = name
 	c.Args = args
 
+	// This inner cancel seems redundant but it doesn't notify upward.
+	// This allows upper lifecycles to finish.
+	ctx, c.tmpCancel = context.WithCancel(ctx)
+
 	cmd := c.makeExecCmd(ctx)
-	err := c.doCapture(cmd)
+	p, err := c.doStart(cmd)
+	if err != nil {
+		return p, err
+	}
+
+	c.tmpCmd = cmd
+
+	return p, nil
+}
+
+func (c *Capture) doStart(cmd *exec.Cmd) (*Pipes, error) {
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create stdout, stderr streams of type io.Reader
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, err
+	}
 
-	return c, err
+	if err = cmd.Start(); err != nil {
+		c.ExitCode = errToExitCode(err)
+	}
+
+	return &Pipes{
+		Stdin:  stdin,
+		Stdout: stdout,
+		Stderr: stderr,
+	}, nil
+}
+
+// Restart runs a previously run command, binding pipes before operation.
+func (c *Capture) Restart(ctx context.Context) (*Pipes, error) {
+	return c.Start(ctx, c.Name, c.Args...)
+}
+
+// Stop ends a Started capture.
+func (c *Capture) Stop() error {
+	if c.tmpCancel == nil || c.tmpCmd == nil {
+		return errors.New("command was not started")
+	}
+
+	err := c.doStop()
+
+	c.ExitCode = errToExitCode(err)
+
+	return err
+}
+
+func (c *Capture) doStop() error {
+	if c.tmpCancel != nil {
+		c.tmpCancel()
+	}
+
+	timeout := time.NewTimer(time.Second * 10)
+	ticker := time.NewTicker(time.Second)
+
+	if c.tmpCmd == nil {
+		return errors.New("wasn't running")
+	}
+
+	if c.tmpCmd.ProcessState == nil || c.tmpCmd.ProcessState.Exited() {
+		if c.tmpCmd.ProcessState != nil {
+			c.ExitCode = c.tmpCmd.ProcessState.ExitCode()
+		}
+
+		c.tmpCmd = nil
+
+		return nil // process is stopped
+	}
+
+	if c.tmpCmd.ProcessState.Exited() {
+		return nil
+	} else {
+		err := c.tmpCmd.Process.Kill()
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case <-timeout.C:
+			ticker.Stop()
+
+			return errors.New("timeout reached")
+		case <-ticker.C:
+			if !c.tmpCmd.ProcessState.Exited() {
+				ticker.Stop()
+				timeout.Stop()
+
+				continue
+			}
+
+			return c.tmpCmd.Wait()
+		}
+	}
 }
 
 // Rerun re-runs the Capture's parameters in a new shell, recording
 // A result to it in the same way as Run.
-func (c Capture) Rerun(ctx context.Context) (Capture, error) {
+func (c *Capture) Rerun(ctx context.Context) (*Pipes, error) {
 	return c.Run(ctx, c.Name, c.Args...)
 }
 
-func (c Capture) makeExecCmd(ctx context.Context) *exec.Cmd {
+func (c *Capture) makeExecCmd(ctx context.Context) *exec.Cmd {
 	cmd := exec.CommandContext(ctx, c.Name, c.Args...)
 	cmd.Env = c.Env
 	cmd.Dir = c.Dir
 
 	return cmd
 }
 
-func (c *Capture) doCapture(cmd *exec.Cmd) (err error) {
-	output, err := cmd.CombinedOutput()
+func (c *Capture) doRun(cmd *exec.Cmd) (*Pipes, error) {
```

Comment:
As written, process I/O is unconfigurable.  I think we can refine the approach to allow for configuring pipes.  

It would only require a `Configure` func and functional options.  Something like, say:

```
type pipesOption func(p *Pipes) pipesOption

func WithStdIn(in *io.WriteCloser) {
  return func(p *Pipes) pipeOption {
      p.Stdin = in
  }

  // same for StdOut & StdErr

  func (p *Pipes) Configure(opts ...pipesOption) {
      for _, opt := range opts {
            opt(p)
      }
  }
}
```

The trick is you need to move where the `Pipes` get initialized, or offer some way to pass down those functional opts to this point.  Since this is a new, different kind of variadic parameter, it doesn't fit well as something to slap into `New()`, since you can't have two different kinds of variadic params as input.  So you'd have to have a `NewPipes(o ...pipesOption)` constructor and pass in `*Pipes` as a parameter to `New()` [also adding another field to `Capture`, like `pipes *Pipes` or something where, if nil, you do the same thing in `doStart` you're doing now].

This is a _suggestion_.  Consider it.  These modifications would be a minor increase in complexity but yield a huge improvement in utility as the package would be more broadly applicable across more generalized cases than it is now.

Reply:
I will consider it, but keep in mind that the properties of the pipes are to deliver to the user the standard directional pipes. You could build a proxy type around Pipes to read/write to whatever you choose, and I even do something like this in rcmd.

Reply:
That's a great point, too.

Reply:
There's also a lot of trickiness that I seek to eliminate here. In Run, we allow the user to peg things to the input, which we define as a `bytes.Buffer` internally. They could, likewise wrap that in another pipe and route as they choose.

Reply:
The question remains whether making this usable is possible, given the situations in which the pipes, defined inputs, etc are used.

Reply:
I guess we employ our old friend try-it-for-a-while-and-see.